File: Python/11-Math&Geometry/Algorithms/1-EuclideanAlgorithm(GCD)/EuclideanAlgorithm(GCD).py
def gcd_recursive(a, b):
    """
    Recursive implementation of the Euclidean Algorithm.

    Args:
        a: The first integer.
        b: The second integer.

    Returns:
        The greatest common divisor of a and b.

    Example:
        gcd_recursive(48, 18) == 6
    """
    if b == 0:
        return a
    else:
        return gcd_recursive(b, a % b)

def gcd_iterative(a, b):
    """
    Iterative implementation of the Euclidean Algorithm.

    Args:
        a: The first integer.
        b: The second integer.

    Returns:
        The greatest common divisor of a and b.

    Example:
        gcd_iterative(48, 18) == 6
    """
    while b:
        a, b = b, a % b
    return a

def gcd_subtraction(a, b):
    """
    GCD calculation using repeated subtraction.

    Args:
        a: The first integer.
        b: The second integer.

    Returns:
        The greatest common divisor of a and b.

    Example:
        gcd_subtraction(48, 18) == 6
    """
    while a != b:
        if a > b:
            a -= b
        else:
            b -= a
    return a

def gcd_binary_recursive(a, b):
    """
    Recursive implementation of the Binary (Stein's) Algorithm.
    This is often faster than the Euclidean algorithm, especially for large numbers.

    Args:
        a: The first integer.
        b: The second integer.

    Returns:
        The greatest common divisor of a and b.

    Example:
        gcd_binary_recursive(48, 18) == 6
    """
    if a == 0:
        return b
    if b == 0:
        return a
    if a == b:
        return a

    if (a % 2 == 0) and (b % 2 == 0):  # Both even
        return 2 * gcd_binary_recursive(a // 2, b // 2)
    elif (a % 2 == 0):  # a is even, b is odd
        return gcd_binary_recursive(a // 2, b)
    elif (b % 2 == 0):  # a is odd, b is even
        return gcd_binary_recursive(a, b // 2)
    elif a > b:
        return gcd_binary_recursive((a - b) // 2, b)
    else:
        return gcd_binary_recursive((b - a) // 2, a)

def gcd_extended(a, b):
    """
    Extended Euclidean Algorithm.  Finds GCD of a and b,
    and also finds x and y such that ax + by = gcd(a, b)

    Args:
        a: The first integer.
        b: The second integer.

    Returns:
        A tuple (gcd, x, y) where gcd is the GCD of a and b,
        and x and y satisfy the equation ax + by = gcd.

    Example:
        gcd_extended(48, 18) == (6, -1, 3)  # 48*(-1) + 18*(3) = 6
    """
    if b == 0:
        return a, 1, 0
    else:
        gcd, x, y = gcd_extended(b, a % b)
        return gcd, y, x - (a // b) * y

# Example usage and output:
a = 48
b = 18

print(f"GCD({a}, {b}) using recursion:", gcd_recursive(a, b))
print(f"GCD({a}, {b}) using iteration:", gcd_iterative(a, b))
print(f"GCD({a}, {b}) using subtraction:", gcd_subtraction(a, b))
print(f"GCD({a}, {b}) using binary recursion:", gcd_binary_recursive(a, b))

gcd, x, y = gcd_extended(a, b)
print(f"GCD({a}, {b}) using extended algorithm:", gcd)
print(f"Extended GCD: {a}*{x} + {b}*{y} = {gcd}")

File: Java/11-Math&Geometry/Algorithms/1-EuclideanAlgorithm(GCD)/EuclideanAlgorithm(GCD).java
import java.util.InputMismatchException;
import java.util.Scanner;

public class EuclideanGCD {

    /**
     * Main method to demonstrate and test the different GCD implementations.
     *
     * @param args command line arguments (not used in this example)
     */
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int num1, num2;

        // Get input from the user with input validation.
        try {
            System.out.print("Enter the first non-negative integer: ");
            num1 = scanner.nextInt();
            if (num1 < 0) {
                System.out.println("Error: Input must be a non-negative integer.");
                scanner.close();
                return; // Exit the program
            }

            System.out.print("Enter the second non-negative integer: ");
            num2 = scanner.nextInt();
            if (num2 < 0) {
                System.out.println("Error: Input must be a non-negative integer.");
                scanner.close();
                return; // Exit the program
            }
        } catch (InputMismatchException e) {
            System.out.println("Error: Invalid input. Please enter integers only.");
            scanner.close();
            return; // Exit the program
        } finally {
            // Close the scanner to prevent resource leaks
            // moved scanner.close() to finally block
            // scanner.close(); // Close the scanner
        }


        // Calculate and print GCD using different methods.
        System.out.println("GCD(" + num1 + ", " + num2 + ") using recursion: " + gcdRecursive(num1, num2));
        System.out.println("GCD(" + num1 + ", " + num2 + ") using iteration: " + gcdIterative(num1, num2));
        System.out.println("GCD(" + num1 + ", " + num2 + ") using binary recursion: " + gcdBinaryRecursive(num1, num2));
        System.out.println("GCD(" + num1 + ", " + num2 + ") using binary iteration: " + gcdBinaryIterative(num1, num2));
        System.out.println("GCD(" + num1 + ", " + num2 + ") using subtraction: " + gcdSubtraction(num1, num2));
        scanner.close(); // moved scanner close here
    }

    /**
     * Recursive implementation of the Euclidean Algorithm.
     * This method calculates the GCD of two numbers using recursion.
     *
     * @param a the first non-negative integer
     * @param b the second non-negative integer
     * @return the greatest common divisor of a and b
     */
    public static int gcdRecursive(int a, int b) {
        if (b == 0) {
            return a;
        }
        return gcdRecursive(b, a % b);
    }

    /**
     * Iterative implementation of the Euclidean Algorithm.
     * This method calculates the GCD of two numbers using a while loop.
     *
     * @param a the first non-negative integer
     * @param b the second non-negative integer
     * @return the greatest common divisor of a and b
     */
    public static int gcdIterative(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    /**
     * Recursive implementation of the Binary (Stein's) GCD Algorithm.
     * This method is more efficient than the traditional Euclidean algorithm, especially for large numbers.
     * It leverages bitwise operations and recursion.
     *
     * @param a the first non-negative integer
     * @param b the second non-negative integer
     * @return the greatest common divisor of a and b
     */
    public static int gcdBinaryRecursive(int a, int b) {
        if (a == 0) {
            return b;
        }
        if (b == 0) {
            return a;
        }
        if (a == b) {
            return a;
        }
        if ((a & 1) == 0) { // a is even
            if ((b & 1) == 0) { // b is even
                return gcdBinaryRecursive(a >> 1, b >> 1) << 1; // result * 2
            } else { // b is odd
                return gcdBinaryRecursive(a >> 1, b);
            }
        } else { // a is odd
            if ((b & 1) == 0) { // b is even
                return gcdBinaryRecursive(a, b >> 1);
            } else { // b is odd
                if (a > b) {
                    return gcdBinaryRecursive(a - b, b);
                } else {
                    return gcdBinaryRecursive(b - a, a);
                }
            }
        }
    }

    /**
     * Iterative implementation of the Binary (Stein's) GCD Algorithm.
     * This method calculates the GCD of two numbers without using division,
     * instead using bitwise operations.
     *
     * @param a the first non-negative integer
     * @param b the second non-negative integer
     * @return the greatest common divisor of a and b
     */
    public static int gcdBinaryIterative(int a, int b) {
        // GCD(0, b) is b, GCD(a, 0) is a.
        if (a == 0)
            return b;
        if (b == 0)
            return a;

        //Find common factor of 2
        int shift;
        for (shift = 0; ((a | b) & 1) == 0; ++shift) {
            a >>= 1;
            b >>= 1;
        }

        // Now, a and b are not both even.
        while ((a & 1) == 0)
            a >>= 1;

        // From here, a is always odd.
        do {
            while ((b & 1) == 0)  // Remove all factors of 2 in b
                b >>= 1;
            if (a > b) {
                int t = b;
                b = a;
                a = t;
            }
            b = b - a;
        } while (b != 0);
        return a << shift;
    }

    /**
     * Implementation of the Euclidean Algorithm using subtraction.
     * This method calculates the GCD of two numbers by repeatedly subtracting
     * the smaller number from the larger number until they are equal.
     *
     * @param a the first non-negative integer
     * @param b the second non-negative integer
     * @return the greatest common divisor of a and b
     */
    public static int gcdSubtraction(int a, int b) {
        if (a == 0) return b;
        if (b == 0) return a;

        while (a != b) {
            if (a > b) {
                a = a - b;
            } else {
                b = b - a;
            }
        }
        return a;
    }
}

